### PR TITLE
Pass custom timeout

### PR DESCRIPTION
**Pull request by kanchansenlaskar**
This is the change to incorporate a custom timeout for the test executions.

### CODE REVIEW

File: lisa/testselector.py [hunk 1]
```diff
@@ -125,6 +125,7 @@ def _apply_settings(
     applied_case_data.use_new_environment = (
         applied_case_data.use_new_environment or case_runbook.use_new_environment
     )
+    applied_case_data.timeout = case_runbook.timeout or applied_case_data.timeout
```
Comment (squirrelsc) [inline lisa/testselector.py:128]: When is this change needed? If it needs to extend the timeout, it should take max of both. If it wants to run test cases faster, it should take min of them. They are conflict goals.

For the `or` condition, it takes case_runbook higher priority, for some designed long running test cases, they will mostly failed. It's not right to overwrite case settings always.

File: lisa/testsuite.py [hunk 2]
```diff
@@ -630,7 +631,10 @@ def start(
                 constants.RUN_LOCAL_LOG_PATH
             ).as_posix()
             case_result.set_status(TestStatus.RUNNING, "")
-            case_timeout = case_result.runtime_data.metadata.timeout
+            case_timeout = (
```
Comment (squirrelsc) [inline lisa/testsuite.py:636]: When is this change needed? If it needs to extend the timeout, it should take max of both. If it wants to run test cases faster, it should take min of them. They are conflict goals.

For the `or` condition, it takes case_runbook higher priority, for some designed long running test cases, they will mostly failed. It's not right to overwrite case settings always.

File: lisa/variable.py [hunk 3]
```diff
@@ -371,6 +371,7 @@ def _replace_variables(data: Any, variables: Dict[str, VariableEntry]) -> Any:
                             f"cannot find variable '{variable_name[2:-1]}', "
                             "make sure its value filled in runbook, "
                             "command line or environment variables."
+                            f"data: {data}"
```
Comment (squirrelsc) [inline lisa/variable.py]: When is it needed? I'm worry about if it prints too early to leak secrets in the log.

File: lisa/testsuite.py (hunk 2 above)
Comment (kanchansenlaskar) [reply lisa/testsuite.py:636]: Changed it, it will take the max of both.
However, it will consider the one in metadata if the one passed from outside is None

File: lisa/testselector.py (hunk 1 above)
Comment (squirrelsc) [inline lisa/testselector.py]: It needs to use max value too.

File: lisa/schema.py [hunk 4]
```diff
@@ -1487,6 +1487,7 @@ class TestCase(BaseTestCaseFilter):
             field_function=fields.Int, validate=validate.Range(min=0)
         ),
     )
+    timeout: int = 3600
```
Comment (squirrelsc) [inline lisa/schema.py]: It needs to be optional, so if some test case is shorter timeout, it should be respected.

Comment (kanchansenlaskar) [reply lisa/schema.py]: made it optional

File: lisa/testselector.py (hunk 1 above)
Comment (kanchansenlaskar) [reply lisa/testselector.py]: done

File: lisa/schema.py [hunk 5]
```diff
@@ -1487,6 +1487,8 @@ class TestCase(BaseTestCaseFilter):
             field_function=fields.Int, validate=validate.Range(min=0)
         ),
     )
+    # each test case will have this timeout, default is 3600 seconds
+    timeout: Optional[int] = 3600
```
Comment (squirrelsc) [inline lisa/schema.py]: It's not optional by default, please set the default value to `None`. So the test case default timeout are respected.

File: lisa/schema.py [hunk 6]
```diff
@@ -1487,6 +1487,9 @@ class TestCase(BaseTestCaseFilter):
             field_function=fields.Int, validate=validate.Range(min=0)
         ),
     )
+    # each test case will have this timeout
+    # if it iss 0 it will use default timeout from test case
```
Comment (squirrelsc) [inline lisa/schema.py]: typo: `iss` to `is`

File: lisa/schema.py (hunk 5 above)
Comment (kanchansenlaskar) [reply lisa/schema.py]: I have set the default value to 0 to maintain the consistency with the other places in other repo. Please do let me know if you feel otherwise.

**Comment by squirrelsc** [comment]
@LiliDeng LGTM. @kanchansenlaskar, please wait Lili having a regression test.

**Review by LiliDeng — APPROVED** [review]
